doc/tutorials/ensemble.md

```diff
@@ -0,0 +1,465 @@
+
+$ gxensemble help lookup
+Usage: gxensemble lookup [command-option ...] <server-or-role>
+       looks up a server by id or role
+
+Command Options:
+  --registry <registry>           additional registry addresses; by default the registry is reachable at unix /tmp/ensemble/registry [default: #f]
+  --role                          lookup by role
+
+Arguments:
+ server-or-role                   the server or role to lookup
+
+$ gxensemble help shutdown
+Usage: gxensemble shutdown [command-option ...] [<server-id>] [<actor-id>]
+       shuts down an actor, server, or the entire ensemble including the registry
+
+Command Options:
+ -f --force                       force the action
+  --registry <registry>           additional registry addresses; by default the registry is reachable at unix /tmp/ensemble/registry [default: #f]
+
+Arguments:
+ server-id                        the server id [default: #f]
+ actor-id                         the actor's registered name [default: #f]
+```
+
+## A Working example: httpd with dynamic handler registration
+
+We can make all this concrete with a working example: an httpd server
+ensemble, that supports dynamic handler registration.
+
+The source code for the tutorial is available in the gerbil source distribution, in
+[src/tutorial/ensemble](https://github.com/vyzo/gerbil/tree/master/src/tutorial/ensemble).
+You can build it using the [build script](https://github.com/vyzo/gerbil/tree/master/src/tutorial/ensemble/build.ss).
+
+Firt let's look at the server implementation in [src/tutorial/ensemble/server.ss](https://github.com/vyzo/gerbil/tree/master/src/tutorial/ensemble/server.ss):
+```scheme
+(import :gerbil/gambit/threads
+        :std/net/httpd
+        :std/io)
+(export run!)
+
+(def (run! port)
+  (let* ((addr (cons inaddr-any4 port))
+         (httpd (start-http-server! addr)))
+    (thread-join! httpd)))
+```
+
+This is a very thin server, that gets started with the `run!` method
+of the module; it takes a port, starts an httpd server and waits for
+it to shutdown.  Notice that the server is initialized with the
+default mux, and it is initially empty; there are no handlers, so
+everything will 404.
+
+We also have a server wrapper for running the server as a service with `gxensemble run`.
+This is the code for the wrapper (see [src/tutorial/ensemble/httpd-svc.ss](https://github.com/vyzo/gerbil/tree/master/src/tutorial/ensemble/httpd-svc.ss)):
+```scheme
+(import ./server)
+(export main)
+
+(def (main (port "8080"))
+  (run! (string->number port)))
+```
+
+With all this, let's start an ensemble with two httpds, named `httpd1` and `httpd2`:
+```
+# generate a cookie for our ensemble, if on does not already exist
+$ gxensemble cookie
+
+# start the registry
+$ gxensemble registry
+...
+
+# start the servers
+$ gxensemble run --roles "(httpd)" httpd1 :tutorial/ensemble/httpd-svc 8080
+...
+
+$ gxensemble run --roles "(httpd)" httpd2 :tutorial/ensemble/httpd-svc 8081
```

Comment:
sure, but then we'd want to use TCP for this and as we discussed I don't want to get people of using TCP instead of TLS and might even drop support for TCP altogether when TLS is ready.

however, with TLS I can make the same demo with two separate machines, that would be neat.